### DISCUSSION TITLE
Generate PDFs from Supabase directly.

### DISCUSSION THREAD

**Discussion by aeynaud** [supabase/supabase#19824] [Questions]
Hi there, 

Apologies if this is not the best place to ask this.
I have JSON data in one of my tables on Supabase, which I would love to be able to turn into a PDF which then gets stored in a Supabase Storage Bucket.

Looking online there are lots of libraries for server-side PDF generation using Node.js (or Next.js?) but my backend is just Supabase.

I also saw that Adobe has a REST API for PDF generation (server-side) but I am not sure how to get to that. Ideally, my goal would be to schedule an Edge function monthly to generate quotes for all my clients.

Could anyone please kindly steer me in the right direction? How can I achieve something like this if I only have Supabase?
Many thanks in advance.

**Comment by GaryAustin1** [comment]
You would need to use an edge function and a pdf library.

**Reply by filippobrigati** [reply]
Hi, i'm trying to return a pdf from an edge function but i get this error

`{"msg":"WorkerRequestCancelled: request has been cancelled by supervisor"}`

If i try to fetch the function from postman all looks good and the pdf file are returned correcty, is this problem related to [this issue](https://github.com/supabase/supabase/issues/18175)?

Thanks

**Comment by Usprime** [comment]
> You would need to use an edge function and a pdf library.  

Hi could you advice which library to use for pdf export?

**Reply by GaryAustin1** [reply]
I used pdfmake http://pdfmake.org/#/

I've used it in an edge function to format passed data in its format with images/tables/etc. and return pdf rawdata from invoke.  I believe the same could be done to return a file from the edge function OR have the edge function store the data in the background to the storage api.  There is a supabase example of storing filedata to storage.

I never finished the coding, but did the basic work to  make sure it could take formatted data and generate a pdf.
